File: convexhull-giftwrapping.py
import math
from functools import cmp_to_key

def solve(points):
    perimeter = 0
    leftMostPoint = findLeftMostPoint(points) 
    firstPoint = leftMostPoint
    nextPoint = None
    while nextPoint != leftMostPoint:
        nextPoint = findNextPoint(firstPoint, points)
        distance = dist(firstPoint, nextPoint)
        perimeter += distance
        firstPoint = nextPoint
    return perimeter

def findLeftMostPoint(points):
    return min(points, key=lambda x: x[0])

def findNextPoint(firstPoint, points):
    return max(filter(lambda p: p != firstPoint, points), key=cmp_to_key(findRelativeDirection(firstPoint)))

def dist(p1, p2):
    return math.sqrt((p1[0] - p2[0]) ** 2   + (p1[1] - p2[1]) ** 2)

def findRelativeDirection(point):
    def anon(a, b):
        ax, ay = a[0] - point[0], a[1] - point[1]
        bx, by = b[0] - point[0], b[1] - point[1]

        crossProduct = ax * by - ay * bx

        if crossProduct > 0: return 1
        if crossProduct < 0: return -1
        return 0
    return anon


# test data
d1 = [(1, 1),(2, 5),(3, 3),(5, 3),(3, 2),(2, 2)]
d2 = [(3, 2),(2, 5),(4, 5)]

print(solve(d1))
print(solve(d2))

File: convexhull-giftwrapping.hs
import Text.Printf
import Data.Function
import Data.List

solve :: [(Int, Int)] -> Double
solve points = calculatePerimeter points leftMostPoint leftMostPoint 0
    where
        leftMostPoint = findLeftMostPoint points

calculatePerimeter :: [(Int, Int)] -> (Int, Int) -> (Int, Int) -> Double -> Double
calculatePerimeter points leftMostPoint currentPoint currentPerimeter
  | currentPoint == leftMostPoint && currentPerimeter /= 0 = currentPerimeter
  | otherwise = 
    let nextPoint = findNextPoint currentPoint points
        distance = dist currentPoint nextPoint
      in calculatePerimeter points leftMostPoint nextPoint (currentPerimeter + distance)


findLeftMostPoint :: [(Int, Int)] -> (Int, Int)
findLeftMostPoint = minimumBy (compare `on` fst) 

findNextPoint currentPoint points = maximumBy (findRelativeDirectionTo currentPoint) (filter (/= currentPoint) points)

findRelativeDirectionTo (x,y) (ax,ay) (bx,by)
  | cp > 0 = GT
  | cp < 0 = LT
  | otherwise = EQ
  where
    p1 = (ax - x, ay - y)
    p2 = (bx - x, by - y)
    crossProduct (x1, y1) (x2, y2) = x1 * y2 - y1 * x2
    cp = crossProduct p1 p2

dist :: (Int, Int) -> (Int, Int) -> Double
dist (x1, y1) (x2, y2) = sqrt $ fromIntegral ((x1 - x2) ^ 2 + (y1 - y2) ^ 2)

d1, d2 :: [(Int, Int)]
d1 = [(1, 1),(2, 5),(3, 3),(5, 3),(3, 2),(2, 2)]
d2 = [(3, 2),(2, 5),(4, 5)]

main :: IO ()
main = do
  n <- readLn :: IO Int
  content <- getContents
  let  
    points = map (\[x, y] -> (x, y)). map (map (read::String->Int)). map words. lines $ content
    ans = solve points
  printf "%.1f\n" ans
